1차 pr review 에 대한 수정과 브랜치 dev 변경

yuni02 opened the pull request:
REFACTOR: 
- 불필요한주석 코드 제거 
- else 삭제 
- 변수명 수정
- overload 한 메서드 하나로 통합 Getter 사용하여 중복 메서드 제거

Hyune-c committed:
✅ 테스트 구현 없이 실행시 오류가 발생하도록 수정

yuni02 committed:
[일반야구 lv1]
테스트 항목 2번 , 3번 통과

yuni02 committed:
t4, t5를 위한 익셉션 메세지 생성

yuni02 committed:
Batting class 발생확률이 동일하지 않아서 batting type이 발생하는 로직 변경

yuni02 committed:
t4와 t5 구현을 위한 메서드 추가 (newGame, fourBallCheck)

yuni02 committed:
일반야구 test 5개의 코드 작성 구현

yuni02 committed:
REFACTOR: Batting.java else 사용하지 않는 코드로 수정

yuni02 committed:
REFACTOR:
1) bPlaying, totOut 변수명 수정
2) overload 한 newGame 매개변수 추가해서 하나로 통합함

yuni02 committed:
REFACTOR: 불필요한 주석 코드 제거

Signed-off-by: yuni02 <strangebride@naver.com>

yuni02 committed:
FIX: @Getter 사용해 mathod 중복 사용 제거

yuni02 committed:
Merge branch 'yunkyeong' into dev

Hyune-c commented on src/main/java/com/hyunec/cosmicbaseballinit/domain/baseball/exception/ExceptionMessage.java:
```diff
@@ -0,0 +1,7 @@
+package com.hyunec.cosmicbaseballinit.domain.baseball.exception;
+
+public class ExceptionMessage {
+        public static final String CANNOT_PROCEED_ATBAT = "현재 타석이 진행중이라, 새로운 게임이 불가능 합니다.";
```
직역하면 `배팅을 진행할 수 없습니다.` 입니다. 새로운 게임과는 의도가 전혀 다릅니다.

yuni02 replied:
넵 직역이 잘못된것 같습니다. 변수명 지을때 조금 더 신중을 기하겠습니다.

Hyune-c commented on src/main/java/com/hyunec/cosmicbaseballinit/domain/baseball/model/Batting.java:
```diff
@@ -3,18 +3,26 @@
-        return Arrays.stream(Batting.values())
-                .filter(e -> e.ordinal() == random.nextInt(2) + 1)
-                .findFirst()
-                .orElse(STRIKE);
+        int nResult = random.nextInt(3) + 1;
+        switch (nResult) {
+            case 1:
+                return STRIKE;
+            case 2:
+                return BALL;
+            default:
+                return HIT;
+        }
     }
```
모두 하드 코딩되어 있네요.
하지만 다음 기능을 하면서 개선될 것으로 기대합니다.

Hyune-c commented on src/main/java/com/hyunec/cosmicbaseballinit/domain/baseball/model/PlateAppearances.java:
```diff
@@ -1,19 +1,59 @@
 @Component
 public class PlateAppearances {
 
     private final List<Batting> battings = new ArrayList<>();
+    private final Map<String, String> player = new HashMap<String, String>();
```
단수형, 복수형 사용에 일관성을 가져주세요.

yuni02 replied:
넵

Hyune-c commented on src/main/java/com/hyunec/cosmicbaseballinit/domain/baseball/model/PlateAppearances.java:
```diff
@@ -1,19 +1,59 @@
     private final List<Batting> battings = new ArrayList<>();
+    private final Map<String, String> player = new HashMap<String, String>();
+
+    private static PlateAppearances pa;
```
이 객체의 의도를 잘 모르겠습니다.
PlateAppearances 안에 왜 또 PlateAppearances 가 있는건가요?

yuni02 replied:
PlateAppearance를 싱글톤으로 하다보니까 변하지 않는 변수가 필요해서 만들었습니다.

Hyune-c replied:
- `@Component` 는 Spring Bean 을 만들어 줍니다.
- 정작 pa 가 사용되는 곳은 `getInstance()` 뿐이고, 그 메서드는 테스트 코드에서만 사용되고 있습니다.
- 도메인 로직과 상태가 한 곳에서 관리되어야만 하나요?

위를 키워드로 다시 생각해봐주세요.

Hyune-c commented on src/main/java/com/hyunec/cosmicbaseballinit/domain/baseball/model/PlateAppearances.java:
```diff
@@ -1,19 +1,59 @@
+    Integer totalOut = 0;
+    boolean isPlaying = false;
+
+    public PlateAppearances() {
+        player.put("1player", "N"); //N 게임 미진행, U: 게임진행중, E: 게임끝
+    }
```
`예상 가능하고 명백하게 구분되는 상태` 라면 String 보다 좋은 방법이 있지 않을까요?

yuni02 replied:
힌트를 주실 수 있을까요? 저번부터 감이 잘 잡 안혀서요

Hyune-c replied:
> key 가 플레이어 이름이고, value 가 게임의 진행 상태인 것을 볼 때, 
player map 에서 관리되는 것 은 `각 플레이어별 게임 진행 상태` 인 것으로 추정됩니다.

- map 의 변수 명이 player 인 것은 적절하지 않습니다. value 가 진행 상태라는 것을 추측할 수 없기 때문입니다.
- 게임의 진행 상태가 명시적으로 3개라면 enum 을 활용할 수 있습니다.

Hyune-c commented on src/main/java/com/hyunec/cosmicbaseballinit/domain/baseball/model/PlateAppearances.java:
```diff
@@ -1,19 +1,59 @@
+//        return this.isPlaying;
+//    }
+
+    public void setPlaying(boolean is) {
+        this.isPlaying = is;
+    }
```
- 변수명을 고민해주세요. 너무 모호합니다.
- setter 의 사용을 다시 고민해주세요.

Hyune-c commented on src/main/java/com/hyunec/cosmicbaseballinit/domain/baseball/model/PlateAppearances.java:
```diff
@@ -1,19 +1,59 @@
+//        return this.totalOut;
+//    }
+
+    public void settotalOutCount() {
+        this.totalOut++;
+    }
```
```suggestion
    public void setTotalOutCount() {
        this.totalOut++;
    }
```

이런 오타는 이력서의 맞춤법이 틀림과 같습니다.
필히 신경 써주세요.

yuni02 replied:
넵 감사합니다. 지적감사드립니다^^

Hyune-c reviewed (changes requested):
전체적으로 setter 의 활용이 너무 잦습니다.
불변성과 객체에 메세지를 전달한다에 대해 좀 더 고민해주세요.

yuni02 committed:
REFACTOR: 변수명 변경

yuni02 committed:
REFACTOR: 싱글콘 구조 제거, 변수명 변경

yuni02 committed:
REFACTOR: 싱글콘 제거에 따른 Test code 변경

yuni02 committed:
REFACTOR: setter 사용 제거

yuni02 reviewed (commented):
코멘트에 대한 코드 수정 후 리뷰요청드립니다.

Hyune-c commented on src/main/java/com/hyunec/cosmicbaseballinit/domain/baseball/model/PlateAppearances.java:
```diff
@@ -40,4 +73,49 @@ public BattingResult result() {
     }
+
+
+    public boolean fourBallCheck() {
+        boolean isFourBallorOut = false;
+        if (!(strikeCount() == 0 && ballCount() == 0)) {
+            if (result().equals(BattingResult.FOUR_BALL) || result().equals(BattingResult.OUT)) {
+                isFourBallorOut = true;
+            }
+        }
+        return isFourBallorOut;
+    }
+
+    public void batting() {
+        this.isPlaying = true;
+
+        if (fourBallCheck()) {
+            totalOut++;
+            battings.clear();
+            this.isPlaying = false;
+
+            //3OUT - NEW CHANGE GAME
+            if (getTotalOut() == 3) {
+                this.totalOut = 0;
+            }
+            return;
+        }
+        batting(Batting.generate());    //4볼이나 OUT이 아니면 리스트에 계속 담는다 > 플레이어 끝
+        batting();
+    }
```
중첩 if 문을 해결해주세요.
early return 만 하여도 많이 해결될 것으로 보입니다.

yuni02 replied:
넵

Hyune-c commented on src/main/java/com/hyunec/cosmicbaseballinit/domain/baseball/model/PlateAppearances.java:
```diff
@@ -40,4 +73,49 @@ public BattingResult result() {
+    }
+
+
+    public String newGame(String str) {
+        if ("force".equals(str)) {
+            totalOut = 0;
+            battings.clear();
+            return null;
+        }
+        if (fourBallCheck()) {
+            totalOut = 0;
+            battings.clear();
+            return ExceptionMessage.NEW_GAME_START;
+        }
+        return ExceptionMessage.CANNOT_PROCEED_NEWGAME;
+    }
```
이 메서드는 newGame 을 하는 init 의 역할만 하면 될 것 같은데, 현재 상태에 대한 판단까지 같이 하고 있네요.
하나의 메서드가 너무 많은 일을 하고 있는 것이 아닌지 고민해주세요.

yuni02 replied:
네 기능을 나눠보겠습니다.

Hyune-c commented on src/test/java/com/hyunec/cosmicbaseballinit/acceptancetest/NormalBaseballLv1Test.java:
```diff
@@ -1,36 +1,140 @@
+        assertEquals(result, true);
+    }
+
+    public static boolean compareThreeValues(int n1, int n2, int n3) {
+        boolean answer = false;
+        boolean equation1 = 0 <= Math.abs(n2 - n1) / 100 && Math.abs(n2 - n1) / 100 < 5;
+        boolean equation2 = 0 <= Math.abs(n3 - n2) / 100 && Math.abs(n3 - n2) / 100 < 5;
+        if (equation1 && equation2) {
+            answer = true;
+        }
+        return answer;
+
     }
```
```suggestion
    public static boolean compareThreeValues(int n1, int n2, int n3) {
        boolean equation1 = 0 <= Math.abs(n2 - n1) / 100 && Math.abs(n2 - n1) / 100 < 5;
        boolean equation2 = 0 <= Math.abs(n3 - n2) / 100 && Math.abs(n3 - n2) / 100 < 5;
        return equation1 && equation2;
    }
```

`nit`

yuni02 replied:
훨씬 간결해보이네요

Hyune-c commented on src/test/java/com/hyunec/cosmicbaseballinit/acceptancetest/NormalBaseballLv1Test.java:
```diff
@@ -1,36 +1,140 @@
+        );
+
+        boolean result = compareThreeValues((int) n1, (int) n2, (int) n3);
+        assertEquals(result, true);
```
```suggestion
        assertTrue(result);
```
`nit` 저는 일관성을 위해 assertThat 의 활용을 추천합니다.

Hyune-c commented on src/main/java/com/hyunec/cosmicbaseballinit/domain/baseball/model/PlateAppearances.java:
```diff
@@ -40,4 +73,49 @@ public BattingResult result() {
+
+        if (fourBallCheck()) {
+            totalOut++;
+            battings.clear();
```
포볼이 되면 배팅을 초기화하는 것까지 같이 하고 있군요.
나쁘지 않은 구현이지만 하나의 메서드가 해야되는 일이 너무 많은 것이 아닌가 고민해주세요.

Hyune-c reviewed (changes requested):
사용하지 않는 코드는 주석이 아니라 지워주세요.

yuni02 committed:
REFACTOR:
1) early return 문 사용으로 중첩 if 문 제거
2) 주석된 코드 제거
3) newGame 메서드의 복합된 기능을 discriminateNewGame로 분할

yuni02 committed:
REFACTOR:
1) asserEquals 를 assertThat으로 변경하여 테스트 코드의 일관성 확보
2) nit 의견 반영 -> 코드의 간결화

yuni02 committed:
REFACOTOR: 불필요한 java 파일 제거

yuni02 committed:
REFACTOR:불필요한 method 제거

yuni02 committed:
REFACTOR: 빈줄 삭제

yuni02 committed:
REFECTOR: Git Test

Hyune-c commented on src/test/java/com/hyunec/cosmicbaseballinit/acceptancetest/NormalBaseballLv1Test.java:
```diff
@@ -1,36 +1,135 @@
+        pa.batting(Batting.valueOf("BALL"));
+
+        //when, then
+        assertThat(pa.newGame()).isEqualTo(ExceptionMessage.NEW_GAME_START);
```
game 의 결과가 exception 이 아니라 message 라는건 매우 이상합니다.

yuni02 replied:
newGame 의 함수명을 바꾸는게 낫다고 제안해주시는 걸까요?

Hyune-c replied:
아니오. 예외 처리는 String 응답이 아니라 예외 객체를 throw 해야된다는 것을 말하는 것입니다.

- exception handling 에 대해 공부해주세요.

Hyune-c commented on src/main/java/com/hyunec/cosmicbaseballinit/domain/baseball/model/PlateAppearances.java:
```diff
@@ -27,17 +35,56 @@ public Integer ballCount() {
-        battings.clear();
+    public boolean fourBallCheck() {
+        boolean isFourBallorOut = false;
+        if(strikeCount() ==0 && ballCount()==0) return false;
+        if (result().equals(BattingResult.FOUR_BALL) || result().equals(BattingResult.OUT)) return true;
```
- if 이후의 공백 처리가 일관적이지 않습니다.
- == 이후 숫자가 나오기 전 공백이 없습니다.
- if 문의 중괄호가 없습니다.
- 조건문이 2개 이상인 if 문이 1줄로 표현되고 있습니다.

https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style

간단하게 맞출 수 있는 기초적인 컨벤션이 많이 어긋나고 있습니다.
sonarlint 및 컨벤션 포매팅을 활용해주세요.

---

컨벤션 오류는 실무라면 리뷰를 거부를 하고, 코테라면 바로 탈락에 준하는 내용입니다.

yuni02 committed:
REFACTOR:
기초적인 컨벤션과 sonarlint 에 따른 포맷팅 적용

yuni02 committed:
REFACTOR:
newGame()의 결과가 true 면 boolean으로 뱉고, 만약 false라면 예외 뱉는 형식으로 수정

Hyune-c commented on src/main/java/com/hyunec/cosmicbaseballinit/domain/baseball/model/PlateAppearances.java:
```diff
@@ -49,42 +49,50 @@ public BattingResult result() {
-            return discriminateNewGame(true);
-        }
-        return discriminateNewGame(false);
+    public boolean newGame() throws Exception {
+       try{
+           if (fourBallCheck()) {
+               totalOut = 0;
+               battings.clear();
+                return true;
+           }else{
+               Exception e = new Exception("고의로 발생시킴");
+               throw e;
+           }
+       }catch(Exception e) {
+           throw new Exception("현재 타석이 진행중이라, 새로운 게임이 불가능 합니다.");
+       }
     }
```
이전보다 발전되었네요. 👍
하지만 여전히 지금의 코드는 적절하지 않습니다.
`체크드 익셉션 vs 언체크드 익셉션` 에 대해 학습해주세요.

---

하지만 이번 pr 이 너무 길어지고 있고, 충분한 피드백이 되었다고 생각하여 approve 합니다.

yuni02 replied:
네 체크드 익셉션 vs 언체크드 익셉션에 대해서 학습해보겠습니다. 리뷰해주시느라 고생많으셨습니다 ㅠ

Hyune-c reviewed (approved):
수정한 부분의 띄어쓰기 컨벤션이 지켜지지 않고 있습니다.
linter 등을 활용하여 컨벤션을 지켜주세요.
